include/rmm/mr/device/device_memory_resource.hpp

```diff
@@ -40,6 +40,29 @@ namespace mr {
  * in the base class. For example, the base class' `allocate` function may log every allocation, no
  * matter what derived class implementation is used.
  *
+ * The `allocate` and `deallocate` APIs and implementations provide stream-ordered memory
+ * allocation. This allows optimizations such as re-using memory deallocated on the same stream
+ * without the overhead of stream synchronization.
+ *
+ * A call to `allocate(bytes, stream_a)` (on any derived class) returns a pointer that is valid to
+ * use on `stream_a`. Using the memory on a different stream (say `stream_b`) is Undefined Behavior
+ * unless the two streams are first synchronized, for example by using
+ * `cudaStreamSynchronize(stream_a)` or by recording a CUDA event on `stream_a` and then
+ * calling `cudaStreamWaitEvent(stream_b, event)`.
+ *
+ * The stream specified to deallocate() should be a stream on which it is valid to use the
+ * deallocated memory immediately for another allocation. Typically this is the stream on which the
+ * allocation was *last* used before the call to deallocate(). The passed stream may be used
+ * internally by a memory_resource for managing available memory with minimal synchronization, and
```

Comment:
Same as above - does this mean `device_memory_resource`?
```suggestion
 * internally by a `memory_resource` for managing available memory with minimal synchronization, and
```